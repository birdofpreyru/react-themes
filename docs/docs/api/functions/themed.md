# themed()
[themed()]: themed
```tsx
import themed from '@dr.pogodin/react-themes';

// Recommended signature:
function themed<ComponentProps extends ThemeableComponentProps>(
  component: ComponentType<ComponentProps>,
  componentName: string,
  defaultTheme: ComponentProps['theme'],
  options?: ThemedOptions<ComponentProps>,
): ThemedComponent<ComponentProps>;
```

<details>
<summary>Full & decorator signatures of [themed()]</summary>

The full signature of [themed()] function is the following
(essentially, it allows to skip `defaultTheme` argument altogether, and
to provide `options` in its position). It is provided mostly for backward
compatibility, as in practice `defaultTheme` argument is always used.
```tsx
function themed<ComponentProps extends ThemeableComponentProps>(
  component: ComponentType<ComponentProps>,
  componentName: string,

  defaultThemeOrOptions?: ComponentProps['theme']
    | ThemedOptions<ComponentProps>,

  options?: ThemedOptions<ComponentProps>,
): ThemedComponent<ComponentProps>;
```
Also for backward compatibility, this variant of [themed()] function exists,
and allows to use it as a (legacy?) decorator. It is not recommended,
as [Decorators](https://github.com/tc39/proposal-decorators) in JavaScript
have a difficult history, with different variants implemented by different
tools ([Babel](https://babeljs.io/docs/babel-plugin-proposal-decorators)),
and language flavours
([TypeScript](https://www.typescriptlang.org/docs/handbook/decorators.html))
at different times, and because decorators do not provide any unqiue
possibilities that can't be achived without them. If ever they finally become
a standard part of JS/TS eco-system, most probably their support by this library
will have to be revised.
```tsx
function themed<ComponentProps extends ThemeableComponentProps>(
  componentName: string,

  defaultThemeOrOptions?: ComponentProps['theme']
    | ThemedOptions<ComponentProps>,

  options?: ThemedOptions<ComponentProps>,
): ThemedComponentFactory<ComponentProps>;
```
</details>


Registers a themed component under the given name, and with an optional default
theme.

[themed()] can be used as a decorator, or in the following (recommended) way:
```jsx
import themed from '@dr.pogodin/react-themes';
import defaultTheme from './default.scss';

function Component() { ... }

export themed(Component, 'ThemedComponent', defaultTheme);
```

When rendered, your component will receive the composed theme via its
`theme` prop. You will just need to pass the values from `theme` into
the `className` attributes of your component elements, as shown in
the [Getting Started](/docs/tutorial/getting-started#basic-themed-component)
tutorial.

## Generic Parameter
- `ComponentProps` extends [ThemeableComponentProps] &mdash; this type describes
  the props accepted by the original component, wrapped by [themed()] function.

## Arguments
- `component` &mdash; **ComponentType\[ComponentProps\]** &mdash; The base
  component to be themed, its props must compty with the [ThemeableComponentProps]
  interface.

- `componentName` &mdash; **string** &mdash; Themed component name, which should
  be used to provide its context theme via [ThemeProvider].

- `defaultTheme` &mdash; **ComponentProps\['theme'\]** &mdash; Default theme,
  in the form of theme key to CSS classname mapping.

**Optional**

- `options` &mdash; [ThemedOptions] &mdash; Additional options accepted by
  the function.

## Returns

The recommended signature of this function returns themed component:
- [ThemedComponent]**\<ComponentProps\>**

The decorator variant returns instead the component factory, which should be
called right away with a single argument &mdash; the base component to wrap:
- [ThemedComponentFactory]**\<ComponentProps\>**

[ThemeableComponentProps]: /docs/api/types/themeablecomponentprops
[ThemedComponent]: /docs/api/types/themedcomponent
[ThemedComponentFactory]: /docs/api/types/themedcomponentfactory
[ThemedOptions]: /docs/api/types/themedoptions
[ThemeProvider]: /docs/api/components/themeprovider
