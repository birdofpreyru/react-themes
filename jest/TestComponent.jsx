import PT from 'prop-types';

export default function Component({ children, theme }) {
  return (
    <div className={theme.container}>
      <div className={theme.content}>
        { children }
      </div>
    </div>
  );
}

Component.propTypes = {
  children: PT.node,
  theme: PT.shape({
    container: PT.string,
    content: PT.string,
  }).isRequired,
};

Component.defaultProps = {
  children: null,
};
