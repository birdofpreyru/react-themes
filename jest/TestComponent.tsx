import type { FunctionComponent } from 'react';

import type { Theme } from '../src';

export type ComponentTheme = Theme<'container' | 'content'>;

export type ComponentProps = {
  children?: React.ReactNode;
  theme: ComponentTheme;
  goodKey?: string;
};

const Component: FunctionComponent<ComponentProps> = ({
  children,
  goodKey = '',
  theme,
}) => (
  <div className={theme.container}>
    <div className={theme.content}>
      { children }
      { goodKey }
    </div>
  </div>
);

export default Component;
