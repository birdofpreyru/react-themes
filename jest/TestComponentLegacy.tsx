// This is a component accepting legacy themes (Themr library).

export type ComponentThemeT = {
  container?: string;
  content?: string;
};

type ComponentPropsT = {
  children?: React.ReactNode;
  theme: ComponentThemeT;
};

export default function Component({ children, theme }: ComponentPropsT) {
  return (
    <div className={theme.container}>
      <div className={theme.content}>
        { children }
      </div>
    </div>
  );
}

Component.defaultProps = {
  children: null,
};
