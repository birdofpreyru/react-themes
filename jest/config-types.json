{
  "rootDir": "..",
  "runner": "jest-runner-tsd",
  "testMatch": ["**/__tests__/ts-types/**/*.(ts|tsx)"]
}
