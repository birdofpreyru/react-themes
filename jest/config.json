{
  "collectCoverage": true,
  "collectCoverageFrom": ["src/**/*.{js,jsx,ts,tsx}"],
  "coverageDirectory": "__coverage__",
  "rootDir": "..",
  "testPathIgnorePatterns": [
    "/__tests__/ts-types/"
  ],
  "transform": {
    "\\.(j|t)sx?$": "babel-jest"
  }
}
