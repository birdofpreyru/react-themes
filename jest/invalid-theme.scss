/* Example of invalid theme. */

.container {
  color: red;
}

.unexpectedField {
  color: green;
}
