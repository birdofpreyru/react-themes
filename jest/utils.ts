/**
 * A few utils for Jest testing.
 */

/* global expect */

/* eslint-disable import/no-extraneous-dependencies */
import Renderer from 'react-test-renderer';
/* eslint-enable import/no-extraneous-dependencies */

/**
 * Auxiliary wrapper around ReactJS Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the rendered tree.
 */
function render(component: React.ReactElement) {
  return Renderer.create(component).toJSON();
}

/**
 * Renders provided ReactJS component with ReactJS Test Renderer,
 * creates/tests the snapshot, and returns the render.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON render of the component.
 */
export function snapshot(component: React.ReactElement) {
  const res = render(component);
  expect(res).toMatchSnapshot();
  return res;
}

export default null;
