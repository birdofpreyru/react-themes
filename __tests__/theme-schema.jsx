/**
 * Tests for theme schema verifier, and casting.
 */

import React from 'react';

import themed from '../src';

import { snapshot } from '../jest/utils';

import themeA from '../jest/theme-a.scss';
import invalidTheme from '../jest/invalid-theme.scss';
import themeWithExtraStyles from '../jest/theme-with-extra-styles.scss';

function Component({ theme }) {
  return JSON.stringify(theme, null, 2);
}

const ThemedComponent = themed('Component', [
  'container',
  'content',
])(Component);

describe('Theme verification', () => {
  test('No errors with the correct theme', () => {
    const res = ThemedComponent.themeType(
      { theme: themeA },
      'theme',
      'Component',
    );
    expect(res).toBe();
  });

  test('Reports correct errors with invalid theme', () => {
    const res = ThemedComponent.themeType(
      { theme: invalidTheme },
      'theme',
      'Component',
    );
    expect(res.message).toMatchSnapshot();
  });

  test('Theme casting works', () => {
    snapshot((
      <ThemedComponent
        theme={themeWithExtraStyles}
        castTheme
      />
    ));
  });
});
