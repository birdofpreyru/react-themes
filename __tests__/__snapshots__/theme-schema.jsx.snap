// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme verification Reports correct errors with invalid theme 1`] = `
"Theme given to Component has multiple issues:
- Misses adhoc tag classes
- Misses context tag class
- Unexpected theme key unexpectedField"
`;

exports[`Theme verification Theme casting works 1`] = `
"{
  \\"container\\": \\"theme-with-extra-styles__container___o7DyA\\",
  \\"content\\": \\"theme-with-extra-styles__content___3dIhn\\",
  \\"ad\\": \\"theme-with-extra-styles__ad___3OAlJ\\",
  \\"hoc\\": \\"theme-with-extra-styles__hoc___FwFj8\\",
  \\"context\\": \\"theme-with-extra-styles__context___4Fpmq\\"
}"
`;

exports[`Theme verification Theme is missing and is required 1`] = `"Theme is not provided to Component component"`;
